### ISSUE TITLE
NSURLSession error with ServerAuthRootsFile in 0.9.13

### ISSUE THREAD

**Issue by np5** [google/santa#88]
It seems that the use of a certificate signed with a custom CA with ServerAuthRootsFile is not working since 0.9.13.

```
santactl[1764:33988] NSURLSession/NSURLConnection HTTP load failed (kCFStreamErrorDomainSSL, -9802)
HTTP Response: -1200 An SSL error has occurred and a secure connection to the server cannot be made.
Preflight failed, aborting run
```

The same setup is working with 0.9.11 and 0.9.12.

Could it be that it is linked to this commit:
https://github.com/google/macops-MOLAuthenticatingURLSession/commit/c7ad86ccb7f9d21e9ac79218592f3c168490f108

And that line of code:
https://github.com/google/santa/blob/c9ec69b0b5e922eed56db48cf70a07a8e48d6f4c/Source/santactl/Commands/sync/SNTCommandSync.m#L118

**Comment by russellhancox** [comment]
Weird, it seems to work for me.

What's the output of ```santactl sync --debug```

**Comment by apfelwerk** [comment]
Hi Russel,

I can reproduce here as well:

```
ladmins-Mac:~ ladmin$ santactl version
santa-driver    | 0.9.11
santad          | 0.9.11
santactl        | 0.9.11
SantaGUI        | 0.9.11
ladmins-Mac:~ ladmin$ santactl sync --debug
Missing Machine Owner.
Server Trust: /O=Apfelwerk/OU=Zentral/CN=zentral/SHA-1=0613a046a0ce7a9f1433f393a1206d54f120b8aa
Preflight complete
Event upload complete
Rule download complete
Postflight complete
Sync completed successfully
```
once upgraded to 0.9.13

```
ladmins-Mac:~ ladmin$ santactl version
santa-driver    | 0.9.13
santad          | 0.9.13
santactl        | 0.9.13
SantaGUI        | 0.9.13
ladmins-Mac:~ ladmin$ santactl sync --debug
Missing Machine Owner.
Server Trust: /O=Apfelwerk/OU=Zentral/CN=zentral/SHA-1=0613a046a0ce7a9f1433f393a1206d54f120b8aa
2016-09-09 06:05:18.184 santactl[5380:137459] NSURLSession/NSURLConnection HTTP load failed (kCFStreamErrorDomainSSL, -9802)
HTTP Response: -1200 An SSL error has occurred and a secure connection to the server cannot be made.
Preflight failed, aborting run
```

**Comment by russellhancox** [comment]
Interesting - it looks like our test is _working_ because the server we're comparing against isn't self-signed, we're just forcing a reduced list of certs to evaluate against.

Tom's setting up a server with a self-signed cert for testing now.

**Comment by tburgin** [comment]
Once I tested with a self signed cert I see the issue. It is a bug with MOLAuthenticatingURLSession. I have an [issue](https://github.com/google/macops-MOLAuthenticatingURLSession/issues/6) open and will submit a pull request shortly with a fix.

Here is a brief explanation:
Calling `SecTrustSetAnchorCertificates` and `SecTrustEvaluate` on the provided `[NSURLProtectionSpace serverTrust]` in certain situations causes a memory leak. To avoid this a new local serverTrust is created with the same certs and polices within `serverCredentialForProtectionSpace:`. After evaluation the new serverTrust could be destroyed, fixing the leak. These changes were included in the v1.8 release of MOLAuthenticatingURLSession, which is what Santa v0.9.13 uses. However it turns out that `[NSURLProtectionSpace serverTrust]` is evaluated again later on by macOS (Not sure why). Only setting AnchorCertificates for the new serverTrust resulted in self signed certs making it past evaluation and credential creation, but later on when macOS tries to use the un-modified `[NSURLProtectionSpace serverTrust]`, there are no trusted roots and that causes the quite ambiguous error you are experiencing.

**Comment by headmin** [comment]
Hi Tom, 
thanks for quick confirmation and detailed explanation.

**Comment by tburgin** [comment]
Fixed pending 0.9.14 release.
